internal/log/logger.go

```diff
@@ -0,0 +1,154 @@
+		}
+	}
+}
+
+// Error logs a message at ERROR level with the specified message and attributes.  If [err] is not nil,
+// it will be logged in an additional attribute called "err".
+func (l *Logger) Error(err error, msg string, kvs ...any) {
+	rec := getLogRecord(slog.LevelError, msg, kvs...)
+	if err != nil {
+		rec.Add(slog.String("err", err.Error()))
+	}
+	ctx, h := context.Background(), l.logger.Handler()
+	if h.Enabled(ctx, rec.Level) {
+		if err := h.Handle(context.Background(), rec); err != nil {
+			dumpLogHandlerError(err, rec)
+		}
+	}
+}
+
+// getLogRecord builds a [slog.Record] with the provided level and message.  [kvs], if provided, must
+// be a list of tuples where the first item is the string key for the attribute and the second is the
+// value.  These are used to construct [slog.Any] attributes.
+func getLogRecord(lvl slog.Level, msg string, kvs ...any) slog.Record {
+	// skip runtime.Callers(), getLogRecord(), and Info()/Debug()/Error()
+	var pcs [1]uintptr
+	runtime.Callers(3, pcs[:])
+
+	rec := slog.NewRecord(time.Now().UTC(), lvl, msg, pcs[0])
+	if len(kvs) > 0 {
+		attrs := make([]slog.Attr, 0, len(kvs)/2)
+		for i := 0; i < len(kvs); i += 2 {
+			k, ok := kvs[i].(string)
+			if !ok {
+				k = fmt.Sprintf("%v", kvs[i])
+			}
+			attrs = append(attrs, slog.Any(k, kvs[i+1]))
+		}
+		rec.AddAttrs(attrs...)
+	}
+	return rec
+}
+
+// replaceRecordAttributes massages the log record attributes before output
+func replaceRecordAttributes(_ []string, a slog.Attr) slog.Attr {
+	switch a.Key {
+	case slog.SourceKey:
+		// trim "source" down to relative path within this module
+		val := a.Value.Any().(*slog.Source)
+		if idx := strings.Index(val.File, "github.com/CrowdStrike/perseus/"); idx != -1 {
+			val.File = val.File[idx+31:]
+		}
+	default:
+	}
+	return a
+}
+
+var (
+	loadRunningInK8sOnce sync.Once
+	runningInK8s         bool
+)
+
+// inK8S returns a boolean value that indicates whether or not the current process is running inside
+// of Kubernetes.  This is used configure the log output to be either logfmt or JSON.
+func inK8S() bool {
+	loadRunningInK8sOnce.Do(func() {
+		_, runningInK8s = os.LookupEnv("KUBERNETES_SERVICE_HOST")
+	})
+	return runningInK8s
+}
+
+// dumpLogHandlerError is called when invoking the underlying [slog.Handler] returns an error to write
+// the error and additional details to [os.Stderr] in the appropriate format (JSON vs key/value) based
+// on the environment.
+func dumpLogHandlerError(err error, rec slog.Record) {
+	var source string
+	rec.Attrs(func(a slog.Attr) bool {
+		switch a.Key {
+		case slog.SourceKey:
+			src := a.Value.Any().(*slog.Source)
+			source = fmt.Sprintf("%s:%d", src.File, src.Line)
+		default:
+		}
+		return true
+	})
+	// generate a JSON "log" if we're running in k8s, logfmt style k/v pairs otherwise
+	if inK8S() {
+		_, _ = fmt.Fprintf(
+			os.Stderr,
+			`{"time":%q, "level":"ERROR","source":%q,"msg":"failure invoking slog.Handler.Handle()","originalMsg":%q,"error":%q}`,
```

Comment:
It probably won't matter in this case since we're dumping JSON (which k8s will ingest) and the whitespace isn't significant but my own OCD insists that I fix it now. 😬

Reply:
I refactored that code a bit for consistency